File: 3310-remove-methods-from-project/3310-remove-methods-from-project.java
class Solution {
    public List<Integer> remainingMethods(int n, int k, int[][] invocations) {
        List<Integer>[] graph = new ArrayList[n];         // {node -> [list of neighbours]}
        for (int i = 0; i < n; i++) {
            graph[i] = new ArrayList<>();
        }

        for (int[] edge : invocations) {
            graph[edge[0]].add(edge[1]); // {0 -> [2, 1], 1 -> [2], 2 -> [], 3 -> [4]}
        }

        boolean[] suspicious = new boolean[n];
        dfs (k, graph, suspicious);

        for (int[] edge : invocations) {
            int u = edge[0], v = edge[1];
            if (!suspicious[u] && suspicious[v]) { // cannot remove anything return 0...n-1
                List<Integer> ans = new ArrayList<>();
                for (int i = 0; i < n; i++) {
                    ans.add(i);
                }
                return ans;
            }
        }

        List<Integer> ans = new ArrayList<>();
        // return safe nodes
        for (int i = 0; i < n; i++) {
            if (!suspicious[i]) {
                ans.add(i);
            }
        }

        return ans;

    }

    public void dfs (int node, List<Integer>[] graph, boolean[] suspicious) {
        suspicious[node] = true;

        for (int neighbour : graph[node]) {
            if (!suspicious[neighbour]) {
                dfs(neighbour, graph, suspicious);
            }
        }
    }
}

File: 3310-remove-methods-from-project/3310-remove-methods-from-project.go
func remainingMethods(n int, k int, invocations [][]int) []int {
    graph := make([][]int, n)

    for _, edge := range invocations {
        u, v := edge[0], edge[1]
        graph[u] = append(graph[u], v)
    }

    suspicious := make([]bool, n)
    dfs(k, graph, suspicious)

    ans := []int{}

    for _, edge := range invocations {
        u, v := edge[0], edge[1]
        if !suspicious[u] && suspicious[v] { // check for safe -> suspicious
            for i := 0; i < n; i++ { // remove nothing, return 0...n-1
                ans = append(ans, i)
            }
            return ans;
        }
    }


    for i := 0; i < n; i++ {
        if !suspicious[i] {
            ans = append(ans, i)
        }
    }

    return ans
}

func dfs (node int, graph [][]int, suspicious []bool) {
    suspicious[node] = true
    for _, neighbour := range graph[node] {
        if !suspicious[neighbour] {
            dfs (neighbour, graph, suspicious)
        }
    }
}